substrate/client/authority-discovery/src/worker/addr_cache.rs

```diff
@@ -33,16 +50,52 @@ pub(super) struct AddrCache {
 	/// it's not expected that a single `AuthorityId` can have multiple `PeerId`s.
 	authority_id_to_addresses: HashMap<AuthorityId, HashSet<Multiaddr>>,
 	peer_id_to_authority_ids: HashMap<PeerId, HashSet<AuthorityId>>,
+
+	on_change: Option<OnAddrCacheChange>,
+}
+impl AddrCache {
+	/// Clones all but the `on_change` handler (if any).
+	fn clone_content(&self) -> Self {
+		AddrCache {
+			authority_id_to_addresses: self.authority_id_to_addresses.clone(),
+			peer_id_to_authority_ids: self.peer_id_to_authority_ids.clone(),
+			on_change: None,
+		}
+	}
+}
+impl std::fmt::Debug for AddrCache {
+	fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+		f.debug_struct("AddrCache")
+			.field("authority_id_to_addresses", &self.authority_id_to_addresses)
+			.field("peer_id_to_authority_ids", &self.peer_id_to_authority_ids)
+			.finish()
+	}
 }
+impl PartialEq for AddrCache {
+	fn eq(&self, other: &Self) -> bool {
+		self.authority_id_to_addresses == other.authority_id_to_addresses &&
+			self.peer_id_to_authority_ids == other.peer_id_to_authority_ids
+	}
+}
+pub(super) type OnAddrCacheChange =
+	Box<dyn Fn(AddrCache) -> Result<(), Box<dyn fmt::Debug>> + 'static>;
 
 impl AddrCache {
 	pub fn new() -> Self {
 		AddrCache {
 			authority_id_to_addresses: HashMap::new(),
 			peer_id_to_authority_ids: HashMap::new(),
+			on_change: None,
 		}
 	}
 
+	pub fn install_on_change_callback<F>(&mut self, on_change: F)
```

Comment:
I'm torn between setting `on_change` to `None` i ctor and using this mutating method, or if we want to make it non-optional and a parameter of ctor `new`.